services/login_service/app/auth/routes.py

```diff
@@ -0,0 +1,151 @@
+        )
+
+    if get_user_from_bigquery(email):
+        raise HTTPException(status_code=400, detail="User already exists")
+
+    hashed_password = generate_password_hash(password)
+
+    email_sent, code = send_verification_email(email)
+    store_verification_code(
+        email=email, verification_code=code, hashed_password=hashed_password
+    )
+
+    if not email_sent:
+        raise HTTPException(status_code=500, detail="Failed to send verification email")
+
+    token = create_access_token({"sub": email})
+    return {
+        "message": f"User {email} created successfully. Please verify your email.",
+        "access_token": token,
+    }
+
+
+@auth_router.post("/login")
+async def login(data: LoginData):
+    email = data.email
+    password = data.password
+
+    user = get_user_from_bigquery(email)
+    if not user:
+        raise HTTPException(status_code=404, detail="User not found")
+
+    if not check_password_hash(user["password"], password):
+        raise HTTPException(status_code=401, detail="Invalid credentials")
+
+    token = create_access_token({"sub": email})
+    return {"access_token": token, "message": "Login successful"}
+
+
+@auth_router.post("/verify_email")
+async def verify_email(data: VerifyEmailData):
+    email = data.email
+    verification_code = data.verification_code
+
+    if not email or not verification_code:
+        raise HTTPException(
+            status_code=400, detail="Email and verification code are required"
+        )
+
+    client = bigquery.Client()
+
+    query = """
+        SELECT verification_code, hashed_password, created_at
+        FROM `annular-net-436607-t0.Instaspy_DS.verification_codes`
+        WHERE email = @email
+        LIMIT 1
+    """
+    job_config = bigquery.QueryJobConfig(
+        query_parameters=[bigquery.ScalarQueryParameter("email", "STRING", email)]
+    )
+
+    query_job = client.query(query, job_config=job_config)
+    result = query_job.result()
+
+    row = None
+    for row in result:
+        break
+
+    if not row:
+        raise HTTPException(
+            status_code=400, detail="No verification code found for this email"
+        )
+
+    stored_code = row.verification_code
+    created_at = row.created_at
+    hashed_password = row.hashed_password
+    current_time = datetime.now(timezone.utc)
+
+    if stored_code == verification_code:
+        if current_time <= created_at + timedelta(minutes=10):
+            if not insert_user_into_bigquery(email, hashed_password):
+
+                raise HTTPException(status_code=500, detail="Failed to create user")
+
+            # delete_verification_code(email)
+            return {"message": "Email verified successfully!"}
+        else:
+            raise HTTPException(status_code=400, detail="Verification code has expired")
+    else:
+        raise HTTPException(status_code=400, detail="Invalid verification code")
```

Comment:
use guard clauses here to remove nested ifs. In this case is not that bad, but it can go nasty very quickly

https://medium.com/lemon-code/guard-clauses-3bc0cd96a2d3